keps/976-plain-pods/README.md

```diff
@@ -538,39 +537,62 @@ spec:
-gated to ungated.
+The Pod reconciler will run in a "composable" mode: a mode where a Workload is
+composed of multiple objects.
+The `jobframework.Reconciler` will be reworked to accomodate this.
 
-In the Pod event handler, we decrement the counter when we see a transition from having
-the scheduling gate `kueue.x-k8s.io/admission` to not having it.
+After a Workload is admitted, each Pod that owns the workload enters the
+reconciliation loop.
+The reconciliation loop collects all the Pods that are not Failed and constructs
+an in-memory Workload. If there is an existing Workload in the cache and it
+has smaller Pod counters than the in-memory Workload, then it is considered
+unmatching and the Workload is evicted.
 
 In the Pod-group reconciler:
 1. If the Pod is not terminated,
   create a Workload for the pod group if one does not exist.
-2. If the Pod is terminated,
-   - If the Workloald doesn't exist or the workload is finished, remove the finalizer.
-3. ungated_pods_in_client: the number of non-terminated pods in the client that are admitted.
-   We only look at non-terminated pods to allow for terminated pods to be replaced.
-4. ungated_pods = ungated_pods_in_client + expected_ungated_pods. Note that this might temporarily
-   lead to double counting.
-2. For gated pods:
-  - If ungated_pods < admission.count, remove the gate, set nodeSelector, an increase
-    expected_ungated_pods
-  - Else,
-    - If ungated_pods_in_informer < admission.count, we can't admit this Pod now to prevent
-      overbooking, but requeue this Pod for retry.
-    - Else, remove finalizer and delete the Pod, as it's beyond the allowed admission.
+2. If the Pod is terminated and (the Workload doesn't exist or the workload is finished),
+   then remove the finalizer.
+3. Build the in-memory Workload. If its podset counters are greater than the
+  stored Workload, then evict the Workload.
+4. For gated pods:
+  - remove the gate, set nodeSelector
 5. If the number of terminated pods with a finalizer is greater than or equal to the admission
   count, and there are no non-terminated Pods, mark the Workload as Finished and remove the
   finalizers from the Pods.
 
 Note that we are only removing Pod finalizers once the Workload is finished. This is a simple way
 of managing finalizers, but it might lead to too many Pods lingering in etcd for a long time after
-terminated. In a future version, we can consider a better scheme similar to [Pod tracking in Jobs](https://kubernetes.io/blog/2022/12/29/scalable-job-tracking-ga/).
+terminated.
+
+
+### Dynamically reclaiming Quota
+
+Succeeded Pods will note be considered replaceable. In other words, the quota
+from Succeeded Pods will be released by filling [reclaimablePods](https://kueue.sigs.k8s.io/docs/concepts/workload/#dynamic-reclaim)
+in the Workload status.
+
+### Retrying Failed Pods
+
+The Pod group will generally only be considered finished if all the Pods finish
```

Comment:
What about pod groups that do not replace failed pods? We have this use case where a user does not allow retries (and I imagine is a common batch case). So a pod group would be finished if all pods have exited, regardless of success or failure.

I think one way to do this is for every pod in the group to have the `kueue.x-k8s.io/last-in-group: true` annotation, if "group finished" does not mean the Workload is cleaned up or other running pods in the group are affected.

This seems weird given the name, but would this be the recommendation? Is it worth some kind of alternative annotation configuring this, like `pod-group-mode: Batch` or `pod-group-retry: false`?

Reply:
Just having one pod with `last-in-group` has the semantics you want: The pod group will be considered Finished when there are no more Running Pods and there is at least one pod with the annotation.

But I agree that the name is maybe not the best for this use case. I'm thinking of alternatives.

Definitely `pod-group-mode: Batch` is not accurate, as batch doesn't imply that retries are not possible.

`pod-group-retry` doesn't fit the semantics I originally wanted that well.

Reply:
I decided to go with `retriable-in-group: false`, similar to your proposal.
But I also added another mechanism for termination: just delete the Workload.